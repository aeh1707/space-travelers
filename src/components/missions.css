table,
tr,
td {
  border: 1px solid gainsboro;
}

table {
  border-collapse: collapse;
  margin: 5%;
}

.bold {
  font-weight: bold;
  padding: 1%;
  width: 10%;
}

td {
  padding: 1%;
}

.desc {
  width: 60%;
}

.buttons {
  width: 12%;
}

.join {
    background: none;
    padding: 5%;
    width: 90%;
}

.leave {
    background: none;
    width: 90%;
    padding: 5%;
    border: 2px solid crimson;
}