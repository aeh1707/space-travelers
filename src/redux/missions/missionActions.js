import { createAsyncThunk } from '@reduxjs/toolkit';

const GET_MISSIONS = 'GET_MISSIONS';
export const RESERVE = 'RESERVE';
export const LEAVE = 'LEAVE';

const fetchMissions = createAsyncThunk(
  GET_MISSIONS, async () => {
    const response = await fetch('https://api.spacexdata.com/v3/missions');
    const data = await response.json();
    const dataArr = [];
    data.forEach((mission) => {
      dataArr.push({
        mission_id: mission.mission_id,
        mission_name: mission.mission_name,
        description: mission.description,
        reserved: false,
      });
    });
    return dataArr;
  },
);

export const reserveMission = (id) => ({ type: RESERVE, payload: id });
export const leaveMission = (id) => ({ type: LEAVE, payload: id });

export default fetchMissions;
