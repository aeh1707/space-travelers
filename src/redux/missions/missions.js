import { createAsyncThunk } from '@reduxjs/toolkit';

const GET_MISSIONS = 'GET_MISSIONS';

const missionReducer = (state = [], action) => {
  switch (action.type) {
    case 'GET_MISSIONS/fulfilled':
      return action.payload;
    default:
      return state;
  }
};

export const fetchMissions = createAsyncThunk(
  GET_MISSIONS, async () => {
    const response = await fetch('https://api.spacexdata.com/v3/missions');
    const data = await response.json();
    const dataArr = [];
    data.forEach((mission) => {
      dataArr.push({
        mission_id: mission.mission_id,
        mission_name: mission.mission_name,
        description: mission.description,
        joined: false,
      });
    });
    return dataArr;
  },
);

export default missionReducer;
