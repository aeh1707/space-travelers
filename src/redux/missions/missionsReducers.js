const missionReducer = (state = [], action) => {
  switch (action.type) {
    case 'GET_MISSIONS/fulfilled':
      return action.payload;
    default:
      return state;
  }
};

export default missionReducer;
