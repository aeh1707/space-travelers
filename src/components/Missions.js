import './missions.css';
import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import fetchMissions, { reserveMission, leaveMission } from '../redux/missions/missionActions';

const Missions = () => {
  const missions = useSelector((state) => state.missions);
  const dispatch = useDispatch();
  useEffect(() => {
    if (!missions.length) {
      dispatch(fetchMissions());
    }
  }, []);
  const handleReserve = (id) => {
    dispatch(reserveMission(id));
  };
  const handleLeave = (id) => {
    dispatch(leaveMission(id));
  };
  return (
    <>
      <table>
        <thead className="bold">
          <tr>
            <td>Mission</td>
            <td>Description</td>
            <td>Status</td>
          </tr>
        </thead>
        <tbody>
          {missions && missions.map((mission) => (
            <tr key={mission.mission_id}>
              <td className="bold">{mission.mission_name}</td>
              <td className="desc">{mission.description}</td>
              <td className="buttons">
                {!mission.reserved && <span className="notMember generalM"> NOT A MEMBER</span> }
                {' '}
                {mission.reserved && <span className="member generalM"> Active Member</span> }
              </td>
              <td className="buttons">
                {!mission.reserved && <button type="button" onClick={() => handleReserve(mission.mission_id)} className="join padding-bg">Join Mission</button>}
                {mission.reserved && <button type="button" onClick={() => handleLeave(mission.mission_id)} className="leave padding-bg">Leave Mission</button>}
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </>
  );
};

export default Missions;
