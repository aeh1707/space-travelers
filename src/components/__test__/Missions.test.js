import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { reserveMission } from '../../redux/missions/missionActions';
import missionReducer from '../../redux/missions/missionsReducers';
import store from '../../redux/configureStore';
import Missions from '../Missions';

describe('To test Mission components', () => {
  test('To Test Missions component renders successfully!', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Missions />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('To test join mission status', () => {
    const Mission = [
      {
        mission_id: 1, mission_name: 'Teslar', description: 'Telstar 19V (Telstar 19 Vantage)...', reserved: false,
      },
      {
        mission_id: 2, mission_name: 'TThaicom', description: 'Thaicom is the name of a series...', reserved: false,
      },
      {
        mission_id: 3, mission_name: 'JCSAT', description: 'The JSAT constellation is a...', reserved: false,
      },
    ];

    expect(missionReducer(Mission, reserveMission(2))).toEqual(
      [
        {
          mission_id: 1, mission_name: 'Teslar', description: 'Telstar 19V (Telstar 19 Vantage)...', reserved: false,
        },
        {
          mission_id: 2, mission_name: 'TThaicom', description: 'Thaicom is the name of a series...', reserved: true,
        },
        {
          mission_id: 3, mission_name: 'JCSAT', description: 'The JSAT constellation is a...', reserved: false,
        },
      ],
    );
  });
});
