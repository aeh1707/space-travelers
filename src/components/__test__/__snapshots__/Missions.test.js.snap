// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`To test Mission components To Test Missions component renders successfully! 1`] = `
<table>
  <thead
    className="bold"
  >
    <tr>
      <td>
        Mission
      </td>
      <td>
        Description
      </td>
      <td>
        Status
      </td>
    </tr>
  </thead>
  <tbody />
</table>
`;
